src/persistence/history.cpp

```diff
@@ -270,12 +282,16 @@ void History::removeFriendHistory(const ToxPk& friendPk)
                                 "   SELECT id from history "
                                 "   WHERE message_type = 'S' AND chat_id=%1);"
                                 "DELETE FROM history WHERE chat_id=%1; "
-                                "DELETE FROM aliases WHERE owner=%1; "
-                                "DELETE FROM peers WHERE id=%1; "
+                                "DELETE FROM chats WHERE id=%1; "
+                                "DELETE FROM aliases WHERE owner=%2; "
+                                "DELETE FROM authors WHERE id=%2; "
                                 "VACUUM;")
-                            .arg(generatePeerIdString(friendPk));
+                            .arg(generateChatIdString().arg(generateAuthorIdString())),
+                            {friendPk.getByteArray(), friendPk.getByteArray(), friendPk.getByteArray(), friendPk.getByteArray(),
```

Comment:
what's happening here?

Reply:
Yeeeeahh..
We used to use Qt's QString::arg to copy the friend pk's hex string representation into the query in all places at once. Since chats and authors are blobs instead of strings, since tox keys and chat ids are inherently binary anyway, we can't just put the public keys into the string anymore. Instead we're binding the arguments, which is done for each t ime the argument is used.

So Qt's QString::arg is replacing all %1 with `(SELECT id FROM chats WHERE uuid = ?)` from `generateChatIdString`, then for each time it's called we're binding the friend pk repeatedly to the query.

It's definitely not nearly as clean as before in this case :S I still think both binary storage of the keys and binding of argument in SQL queries are conceptually better - but out string generation in History as part of queries is taking a hit.

Reply:
Id probably argue that the efficiency improvements in the db arent worth this change. Hex encoding the strings is 2x the size and adds a little computational cost, but i think everything touched in history is unfortunately worse because of it,. Even though conceptually this is more correct, I think real world impact is negligible and maintencne cost is high. My recommendation is to back out the blob changes

Reply:
To summarize some private discussion:
1) I agree the size/efficiency difference is tiny. Keys in general are a tiiiny portion of our rows compared to chat history, and each one is small.
2) Every other binary field in our db is BLOB - so it's a bit unintuitive from History's perspective to know to hex encode only peers/authors/chats.
3) Agreed that still overall, History's usage is pretty nasty because of this. The injection of bound arguments from the generate strings is surprising, and having to copy the same bound argument 9 times is nasty.
4) I think generally, we should favour argument binding rather than putting arguments in the query string i.e. "SQL injection mode" - so generally would like the move to binding more arguments (maybe even ints in the future).
5) I think the best of both would be if we could have argument binding in history be as clean as QString::arg's usage - i.e. we could just tell RawHistory "put this binary argument at these 9 places" and it just worked. If somehow the generateString helper functions could return both a query section with a bound arg too, and History could bind to many places all at once, that would be nice. I'll take a look at implementing this.

Reply:
I've updated the generate.*String functions to take in the current query string and bound arg vector so that they can add onto both at the point in the query string where they belong. So now the query string and bound args get build together in pieces, making each bound arg much easier to follow where it goes.

It's still probably not as nice as before, but I don't think there's a great way to do similar functionality. Any parsing of the query string in rawdatabase to bind single arguments possibly multiple times seems complex and at risk of SQL injection.

Reply:
This is so much better!